src/Microsoft.AspNet.Mvc.Razor.Host/ViewHierarchyUtility.cs

```diff
@@ -77,18 +78,31 @@ private static IEnumerable<string> GetHierarchicalPath(string relativePath, stri
                 // If the specified path is for the file hierarchy being constructed, then the first file that applies
                 // to it is in a parent directory.
                 relativePath = Path.GetDirectoryName(relativePath);
+
                 if (string.IsNullOrEmpty(relativePath))
                 {
                     return Enumerable.Empty<string>();
                 }
             }
 
+            var builder = new StringBuilder(relativePath);
+            builder.Replace('\\', '/');
+
+            if (builder.Length > 0 && builder[0] != '/')
+            {
+                builder.Insert(0, '/');
+            }
+
             var locations = new List<string>();
-            while (!string.IsNullOrEmpty(relativePath))
+            for (var index = builder.Length - 1; index >= 0; index--)
             {
-                relativePath = Path.GetDirectoryName(relativePath);
-                var path = Path.Combine(relativePath, fileName);
-                locations.Add(path);
+                if (builder[index] == '/')
+                {
+                    builder.Length = index + 1;
+                    builder.Append(fileName);
+
+                    locations.Add(builder.ToString());
```

Comment:
It seemed easy to do it this way rather than doing string concats. I guess there's some perf benefit here but it might become negligible once we figure out some way to cache these.